### ISSUE TITLE
Grouping custom templates in the templates list 

### ISSUE THREAD

**Issue by jameskoster** [WordPress/gutenberg#48651]
Custom templates are somewhat unique as they can be applied _manually_ to multiple post types, rather than _automatically_ according to the template hierarchy. This makes them worthy of some separation in the templates list:

<img width="1560" alt="Templates" src="https://github.com/WordPress/gutenberg/assets/846565/d3350d61-6415-4d86-853f-7ab3df0c003a">

<details><summary>Original issue</summary>There have been a few issues (https://github.com/WordPress/gutenberg/issues/44302, https://github.com/WordPress/gutenberg/issues/42325, https://github.com/WordPress/gutenberg/issues/48589) where the idea of grouping templates has come up. Doing so would enable a couple of practical use cases in the near term:

1. Improve the presentation in the list in the Site Editor's Templates panel (e.g. https://github.com/WordPress/gutenberg/issues/44302#issuecomment-1432901662).
2. Delineate the purpose of some confusing templates like Home
3. Allow us to simplify the Add Template menu which gets very long as you add more plugins

My initial thought would be to use post types as a primary grouping method, with some ad hoc groups for certain core templates. Example:

| Pages | Blog / Posts | General | Custom |
| --- | --- | --- | --- |
| Front Page<br>Page<br>Attachment † <br>Privacy Policy | Home<br>Single Post<br>Category<br>Tag<br>Date<br>Author | Taxonomy<br>Archive<br>Singular<br>Search<br>404<br>Custom | Any custom templates |

† I believe attachments are technically `post`s, but Pages feel like a better fit, in my mind at least.

I think this method can scale quite nicely to accommodate custom post types. It also opens the door to much simpler feature toggling in the future. For example adding a Blog to a site that doesn't already have one could be a single action that adds Home, Single, Category, Tag, Date, Author all at once. By extension it could even be a gateway to a UI for creating custom post types / taxonomies.

Before we can enable any such refinements/actions in the UI, I presume we'll need to install an API, and of course decide on _how_ to group templates. That's what this issue is for. The above is only an initial suggestion to kick off discussion.</details>

**Comment by annezazu** [comment]
Let's both iterate on this from a dev standpoint and continue the discussion here as things evolve! Adding to the needs dev section of phase 2.

**Comment by annezazu** [comment]
This came up as a point of confusion once more in the [FSE Outreach Program's Build a Block Theme exploration](https://make.wordpress.org/test/2023/03/29/fse-program-exploration-build-a-block-theme/#comment-2965):

>I find the list of templates a bit confusing. It should show the most common ones in the top. Less used further down. Here it shows in an alphabetical order. There should be hints along the way to help users get to the correct template.

**Comment by jasmussen** [comment]
What's a good next step to get momentum here? Should we dive into a proof of concept branch? We do know we need the categorization, so perhaps getting a feel for it in a branch is a good start?

Categorizing according to post types seems good to me. For me personally the main thing I'd like to see is not having 404 being sorted at the top :)

**Comment by paaljoachim** [comment]
> Should we dive into a proof of concept branch? We do know we need the categorization, so perhaps getting a feel for it in a branch is a good start?

That sounds like a good idea Joen!

**Comment by paaljoachim** [comment]
Adding in the concept shared by @jameskoster here: 
https://github.com/WordPress/gutenberg/issues/44302#issuecomment-1432901662

<img width="380" alt="219344616-229a2ac4-6049-4459-95f7-66f3e03a0601" src="https://user-images.githubusercontent.com/5323259/231462001-99e1f02c-83f2-4052-ae63-a38ab28dbfca.png">

This is a nice and well structured way to view templates.

**Comment by jameskoster** [comment]
It's a bit tricky because the UI where the groupings will appear (IE the dark material) is somewhat in flux. 

The list in the screenshot above may transition to be "Recent templates" to avoid duplication with the template management view. See the video below for an example, or try it yourself in @SaxonF's [pages-list](https://github.com/WordPress/gutenberg/compare/trunk...try/pages-list) branch. 

https://user-images.githubusercontent.com/846565/231478146-756d2fab-b563-4988-bfb8-a6db5231fa44.mp4

Another candidate would be the Add Template menu, but even that might change now that we have the panels in the dark material, see: https://github.com/WordPress/gutenberg/issues/41060#issuecomment-1486061681

---

I still think the groupings have a place (filtering the management view for example) but it's probably worth aligning on the designs / flows above before committing too much time.

**Comment by jasmussen** [comment]
There seems to be some overlap with #50418 and this one. In light of that, what's a good next step, designwise?

**Comment by jameskoster** [comment]
At the least, I think it would be useful to separate custom templates in the main list:

<img width="1560" alt="Templates" src="https://github.com/WordPress/gutenberg/assets/846565/1e572e94-37c1-4fe2-b972-74144e4140f8">

**Comment by jasmussen** [comment]
Looks good, but the subheading is almost invisible there. Can we add some extra spacing above it to make the separation more clear? I also may be misremembering how we style those subheadings, is it always the same color? I could swear I saw a darker colored version somewhere.

**Comment by jameskoster** [comment]
More spacing seems reasonable 👍 

Regrettably all the colors are still up in the air 🙃 For now the title should match whatever color is used in panels such as Page details. We can always circle back and polish these details after feature-freeze.

**Comment by jasmussen** [comment]
I think you can go big with the spacing though. Scrolling works well and breaks up the long list into chunks.

**Comment by jameskoster** [comment]
I've updated this issue to narrow the scope right down to separating _custom_ templates.

**Comment by jasmussen** [comment]
Let's make it happen!

**Comment by ndiego** [comment]
Just checking in on this one as tomorrow is 6.3 Beta 4. Has any progress been made on this issue, or should we punt to 6.4?

**Comment by jameskoster** [comment]
Probably fine to punt.

**Comment by annezazu** [comment]
I don't agree on punting this. It's a really bad experience that 404 is at the top. I'll try to track down a dev but can leave for now.

**Comment by mtias** [comment]
What happened with the slight title rename so that it'd be Page: 404 or something like that?

**Comment by annezazu** [comment]
Tracked that down and it looks like it should have made it into beta 1 but somehow it doesn't show correctly with beta 3. Commented on the initial PR that handled that change: https://github.com/WordPress/gutenberg/pull/51428#issuecomment-1629378714 

Confirming we can remove this from 6.3 but that we will need to track this down otherwise. It's on my list.

**Comment by jameskoster** [comment]
> I don't agree on punting this. It's a really bad experience that 404 is at the top. I'll try to track down a dev but can leave for now.

Just to be clear, this issue is only about grouping _custom_ templates together.

**Comment by annezazu** [comment]
For context, I was asked to consolidate this prior issue around the 404 template placement here so I think it's a bit of a broader encompassing issue at this point: https://github.com/WordPress/gutenberg/issues/48697#issuecomment-1581731282

**Comment by afercia** [comment]
> this issue is only about grouping custom templates together.

Grouping the custom templates togther and adding a title (heading with proper level) above the group would certainly help also accessibility.

However, that would _de facto_ leave the UI with two groups:
- custom templates, preceeded by a heading
- the other templates, with no heading to explain what they are 

This is not about _visuals_. It's about structuring content in a meaningful and semantic way. If we are going to make two groups, both should have a heading.

A small note: Please remember to mark any issue that touches content structure, headings, and such with the accessibility or the needs accessibility labels. Thank you.

**Comment by jameskoster** [comment]
The other templates are already underneath the main panel heading "Templates". Is that inadequate? 

Custom templates are special group worth separating from regular templates.

**Comment by richtabor** [comment]
An aside here, any reason we shouldn't use drill down here, instead of listing out templates entirely? 

Just thinking about how other plugins can integrate with their templates, like WooCommerce for example, which quickly makes this view quite unmanageable. With their own views, there may be room to help depict the hierarchy as well, especially with all the blog templates (a follow-up exploration). 

![CleanShot 2023-10-31 at 13 37 52](https://github.com/WordPress/gutenberg/assets/1813435/bb0f547e-4421-40b5-9e73-595da99a1d5b)

**Comment by jameskoster** [comment]
I think we explored that before, but the added clicks (both drilling down and up) made it feel a bit awkward. 

There's a balance to strike between this view and Manage templates. At the minute there's quite a bit of overlap given this list is strictly ordered and includes all(?) templates. 

We also have the Grid data view to consider – would it be better if the Templates screen was more like the Patterns one? IE a grid on the right and categories (like you have in your mockup) on the left.

**Comment by mtias** [comment]
I'd like to see an exploration that prioritizes a heading for plugins registering blocks, similar to plugin categories in the block inserter. ideally tied to the APIs we offer for registering templates in the first place. For example, Woo's templates would be under a "Woo" heading; same for plugins like "Sensei" or Newsletter Glue". This is certainly hard to reason about:

<img width="379" alt="image" src="https://github.com/WordPress/gutenberg/assets/548849/e5dec43e-5158-4fa9-8f0c-d996a384bb77">

**Comment by jameskoster** [comment]
I think the design in the OP can handle plugins as you describe. E.g:

<img width="1218" alt="Templates" src="https://github.com/WordPress/gutenberg/assets/846565/e7fcb8dc-352a-444c-b994-5e27b5fa50b2">
